### PR TITLE
Spring security 2

### PR DESCRIPTION
**Pull request by fedaktaras**

### CODE REVIEW

File: src/main/java/com/example/bookstoreapp/exception/CustomGlobalExceptionHandler.java [hunk 1]
```diff
@@ -59,6 +59,16 @@ public ResponseEntity<Object> handleRuntimeException(RuntimeException ex, WebReq
         return new ResponseEntity<>(body, HttpStatus.INTERNAL_SERVER_ERROR);
     }
 
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(RegistrationException ex,
+                                                           WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put(TIMESTAMP_KEY, LocalDateTime.now());
+        body.put(STATUS_KEY, HttpStatus.INTERNAL_SERVER_ERROR);
+        body.put(ERRORS_KEY, "User with such username exists: " + ex.getMessage());
```
Comment (vovasalyha) [inline src/main/java/com/example/bookstoreapp/exception/CustomGlobalExceptionHandler.java]: Not required but it's better to get this message from the exception itself.

File: src/main/java/com/example/bookstoreapp/exception/CustomGlobalExceptionHandler.java [hunk 2]
```diff
@@ -59,6 +59,16 @@ public ResponseEntity<Object> handleRuntimeException(RuntimeException ex, WebReq
...
+        return new ResponseEntity<>(body, HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (vovasalyha) [inline src/main/java/com/example/bookstoreapp/exception/CustomGlobalExceptionHandler.java]: Why 500 status? For client errors we usually respond with 4xx status.

File: src/main/resources/db/changelog/changes/04-create-table-users-roles.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+databaseChangeLog:
+  - changeSet:
+      id: 04
+      author: Taras_Fedak
+      changes:
+        - createTable:
+            tableName: users_roles
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/04-create-table-users-roles.yaml:7]: Make user_id + role_id a composite PK to disallow duplicate entries.

File: src/main/resources/db/changelog/changes/07-insert-ADMIN-to-id-1.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+      changes:
+        - insert:
+            tableName: users_roles
+            columns:
+              - column:
+                  name: user_id
+                  valueComputed: "(SELECT id FROM users WHERE email = 'tarasfedak@me.com')"
+              - column:
+                  name: roles_id
+                  valueComputed: "(SELECT id FROM roles WHERE role = 'ADMIN')"
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/07-insert-ADMIN-to-id-1.yaml]: Don't forget about empty lines.

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Looks good, but see a couple of comments.

File: src/main/resources/db/changelog/changes/07-insert-ADMIN-to-id-1.yaml (hunk 4 above)
Comment (fedaktaras) [reply src/main/resources/db/changelog/changes/07-insert-ADMIN-to-id-1.yaml]: > Don't forget about empty lines.

line 15 added

File: src/main/java/com/example/bookstoreapp/controller/BookController.java [hunk 5]
```diff
@@ -33,18 +34,21 @@ public BookController(BookService bookService) {
     }
 
     @GetMapping
+    @PreAuthorize("hasAnyRole('ROLE_USER','ROLE_ADMIN')")
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapp/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasAnyRole('USER','ADMIN')")
```

Comment (d1sam) [reply src/main/java/com/example/bookstoreapp/controller/BookController.java]: and in another places

File: src/main/java/com/example/bookstoreapp/dto/UserRegistrationRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+@Data
+@FieldMatch(firstFieldName = "password", secondFieldName = "passwordRepeated")
+public class UserRegistrationRequestDto {
+    @NotBlank
+    @Email
+    private String email;
+
+    @NotBlank
+    @Size(min = 8, message = "Password must be at least 8 characters long")
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapp/dto/UserRegistrationRequestDto.java]: what about max?

File: src/main/java/com/example/bookstoreapp/exception/CustomGlobalExceptionHandler.java [hunk 7]
```diff
@@ -55,10 +55,20 @@ public ResponseEntity<Object> handleRuntimeException(RuntimeException ex, WebReq
         Map<String, Object> body = new LinkedHashMap<>();
         body.put(TIMESTAMP_KEY, LocalDateTime.now());
         body.put(STATUS_KEY, HttpStatus.INTERNAL_SERVER_ERROR);
-        body.put(ERRORS_KEY, "An unexpected error has occurred: " + ex.getMessage());
+        body.put(ERRORS_KEY, ex.getMessage());
         return new ResponseEntity<>(body, HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapp/exception/CustomGlobalExceptionHandler.java:59]: please replace `HttpStatus.INTERNAL_SERVER_ERROR`

File: src/main/java/com/example/bookstoreapp/repository/RoleRepository.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package com.example.bookstoreapp.repository;
+
+import com.example.bookstoreapp.model.Role;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapp/repository/RoleRepository.java:9]:
```suggestion
@Repository
public interface RoleRepository extends JpaRepository<Role, Long> {
```

File: src/main/java/com/example/bookstoreapp/repository/UserRepository.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package com.example.bookstoreapp.repository;
+
+import com.example.bookstoreapp.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface UserRepository extends JpaRepository<User, Long> {
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapp/repository/UserRepository.java:9]:
```suggestion
@Repository
public interface UserRepository extends JpaRepository<User, Long> {
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 10]
```diff
@@ -1,3 +1,15 @@
...
+  - include:
+      file: db/changelog/changes/03-create-role-table.yaml
+  - include:
+      file: db/changelog/changes/04-create-table-users-roles.yaml
+  - include:
+      file: db/changelog/changes/05-insert-roles.yaml
+  - include:
+      file: db/changelog/changes/06-create-admin.yaml
+  - include:
+      file: db/changelog/changes/07-insert-ADMIN-to-id-1.yaml
```
Comment (d1sam) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/07-insert-ADMIN-to-id-1.yaml
      
```

File: src/main/resources/db/changelog/changes/07-insert-ADMIN-to-id-1.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/07-insert-ADMIN-to-id-1.yaml:1]: what about preConditions ?

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but there are some changes, that should be done according to my comments.

File: src/main/java/com/example/bookstoreapp/exception/CustomGlobalExceptionHandler.java (hunk 7 above)
Comment (fedaktaras) [reply src/main/java/com/example/bookstoreapp/exception/CustomGlobalExceptionHandler.java:59]: > please replace `HttpStatus.INTERNAL_SERVER_ERROR`

Don't really understand this. With what? It is for RuntimeException. I believe it should be INTERNAL_SERVER_ERROR. Or maybe I don't get smth. Discussed on Q&A.

**Review by andrii-hoienko — APPROVED** [review]
Good job!

**Pull request merged**